File: app/src/main/java/ru/malinoil/geeknote/MainActivity.java
package ru.malinoil.geeknote;

import android.content.res.Configuration;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import ru.malinoil.geeknote.fragments.NoteListFragment;
import ru.malinoil.geeknote.fragments.NotebookFragment;
import ru.malinoil.geeknote.models.NoteEntity;

public class MainActivity extends AppCompatActivity implements NoteListFragment.Controller,
        NotebookFragment.Controller {

    public static final String NOTE_LIST_FRAGMENT_TAG = "NOTE_TAG";

    private BottomNavigationView navigationView;
    private boolean isLand = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.single_fragment_container, new NoteListFragment(), NOTE_LIST_FRAGMENT_TAG)
                .commit();

        isLand = getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
        if (!isLand) {
            renderBottomNavigation();
        }

    }

    @Override
    public void updateNote(NoteEntity noteEntity) {
        if (!isLand) {
            getSupportFragmentManager().popBackStack();
        } else {
            Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.details_container);
            if (fragment != null) {
                getSupportFragmentManager().beginTransaction()
                        .remove(fragment).commit();
            }
        }
        NoteListFragment noteListFragment = (NoteListFragment) getSupportFragmentManager()
                .findFragmentByTag(NOTE_LIST_FRAGMENT_TAG);
        if (noteListFragment != null) {
            noteListFragment.addNoteOnList(noteEntity);
        }
    }

    @Override
    public void openNote(NoteEntity noteEntity) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        if (!isLand) {
            transaction.addToBackStack(null)
                    .replace(R.id.single_fragment_container, NotebookFragment.newInstance(noteEntity));
        } else {
            transaction.add(R.id.details_container, NotebookFragment.newInstance(noteEntity));
        }
        transaction.commit();
    }

    @Override
    public void createNote() {
        openNote(null);
    }

    private void renderBottomNavigation() {
        navigationView = (BottomNavigationView) findViewById(R.id.bottom_navigation);
        navigationView.setOnNavigationItemSelectedListener(item -> {
            FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
            switch (item.getItemId()) {
                case R.id.nav_list:
                    Fragment fragment = getSupportFragmentManager().findFragmentByTag(NOTE_LIST_FRAGMENT_TAG);
                    fragmentTransaction.replace(R.id.single_fragment_container,
                            (fragment == null) ? new NoteListFragment() : fragment);
                    break;
                case R.id.nav_add_note:
                    fragmentTransaction.addToBackStack(null)
                            .replace(R.id.single_fragment_container,
                                    NotebookFragment.newInstance(null));
                    break;
            }
            fragmentTransaction.commit();
            return true;
        });
    }
}

File: app/src/main/java/ru/malinoil/geeknote/MainActivity.kt
package ru.malinoil.geeknote

import android.content.res.Configuration
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import ru.malinoil.geeknote.databinding.ActivityMainBinding
import ru.malinoil.geeknote.models.controllers.FragmentController
import ru.malinoil.geeknote.models.controllers.NoteController
import ru.malinoil.geeknote.ui.fragments.NoteListFragment
import ru.malinoil.geeknote.ui.fragments.NoteFragment
import ru.malinoil.geeknote.models.entities.NoteEntity

class MainActivity : AppCompatActivity(), NoteController, FragmentController {
    private val binding by lazy { ActivityMainBinding.inflate(layoutInflater) }
    private val isLand: Boolean by lazy {
        resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
    }

    companion object {
        const val NOTE_LIST_FRAGMENT_TAG = "NOTE_LIST_TAG"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        setFragment(NoteListFragment(), NOTE_LIST_FRAGMENT_TAG)
        binding.addFab.setOnClickListener {
            openNote(null)
        }
    }

    override fun openNote(note: NoteEntity?) {
        binding.addFab.visibility = View.INVISIBLE
        if(!isLand) {
            setFragment(NoteFragment.newInstance(note), null, isBackStack = true)
        } else {
            setFragment(
                NoteFragment.newInstance(note),
                null, binding.detailsContainer!!.id
            )
        }
    }

    override fun closeNote(note: NoteEntity?) {
        binding.addFab.visibility = View.VISIBLE
        with(supportFragmentManager) {
            if(!isLand) { popBackStack() }
            else {
                findFragmentById(binding.detailsContainer!!.id)?.let { fragment ->
                    beginTransaction().remove(fragment).commit()
                }
            }
        }
        val noteListFragment = (searchFragmentByTag(NOTE_LIST_FRAGMENT_TAG) as NoteListFragment?)
        note?.let {
            noteListFragment?.saveNote(it)
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        binding.addFab.visibility = View.VISIBLE
    }

    override fun setFragment(
        fragment: Fragment,
        tag: String?,
        containerId: Int,
        isBackStack: Boolean)
    {
        val transaction = supportFragmentManager.beginTransaction()
            .replace(containerId, fragment, tag)
        if(isBackStack) {
            transaction.addToBackStack(null)
        }
        transaction.commit()
    }

    override fun searchFragmentByTag(tag: String?): Fragment? {
        return supportFragmentManager.findFragmentByTag(tag)
    }
}